campfire/src/web/serve.rs

```diff
@@ -0,0 +1,249 @@
+
+            let mut needs_rebuild = false;
+            for event in events {
+                match event.event.kind {
+                    EventKind::Create(CreateKind::File) => {
+                        for path in &event.paths {
+                            log::info!("File created: {path:?}");
+                            watcher.add(path)?;
+                        }
+                        needs_rebuild = true;
+                    }
+                    EventKind::Create(CreateKind::Folder) => {
+                        for path in &event.paths {
+                            log::debug!("Folder created: {path:?}");
+
+                            process_results(find_watched_dirs(path), |mut v| {
+                                v.try_for_each(|v| watcher.add(v.path()))
+                            })
+                            .context("Failed to watch new folder")??;
+                        }
+                        needs_rebuild = true;
+                    }
+
+                    EventKind::Modify(v) => {
+                        log::debug!("Modified {v:?}");
+                        needs_rebuild = true;
+                    }
+                    EventKind::Remove(RemoveKind::Folder) => {
+                        for path in &event.paths {
+                            watcher.remove(path)?;
+                        }
+                        needs_rebuild = true;
+                    }
+                    v => {
+                        log::debug!("Other event: {v:?}");
+                    }
+                }
+            }
+
+            if needs_rebuild {
+                log::debug!("Rebuilding...");
+                if let Err(err) = build::run(&self.build).await {
+                    log::error!("Failed to build: {err}")
+                }
+
+                log::debug!("Finished building the web client");
+            }
+        }
+
+        Ok(())
+    }
+}
+
+// pub fn update_watch_subdir(
+//     watching: &mut BTreeSet<PathBuf>,
+//     watcher: impl Watcher,
+//     dir: impl AsRef<Path>,
+// ) -> anyhow::Result<()> {
+//     for entry in find_watched_dirs(dir.as_ref()) {
+//         let entry = entry?;
+
+//         let path = entry.path();
+//         if watching.insert(path.to_path_buf()) {
+//             log::info!("Watching new entry: {path:?}");
+//             watcher.watch(entry.path(), RecursiveMode::NonRecursive)?;
+//         }
+//     }
+
+//     Ok(())
+// }
+
+pub fn find_watched_dirs(
+    dir: impl AsRef<Path>,
+) -> impl Iterator<Item = Result<walkdir::DirEntry, walkdir::Error>> {
+    let dir = dir.as_ref();
+    log::info!("Walking directory {dir:?}");
+
+    walkdir::WalkDir::new(dir).into_iter().filter_entry(|v| {
+        let path = v.path();
+        let fname = v.file_name();
+
+        // if  path.starts_with(".") {
+        //     log::debug!("Ignoring hidden path: {path:?}");
+        //     return false;
+        // }
+
+        match fname.to_str() {
```

Comment:
looking at this - maybe, as a future extension, it should respect `.gitignore`? definitely not necessary now, just figured I'd plant the idea

Reply:
Yes, figured as such as well.

*Oh `.DS_STORE` how I've missed you*